File: motet-sdk/src/motet_sdk/cli/conversations.py
"""
Motet - Conversations CLI

Copyright (c) 2024-2026 Motet Contributors
Licensed under the Apache License, Version 2.0.

Author: Matt Chisholm <matt@motet.dev>
Last Modified: 2026-08-25

Description:
    CLI for conversations — calls /api/v1/conversations.
    List, get, clear, rename, delete.

Dependencies:
    - click: CLI framework
    - motet.cli._auth: get_api_headers
    - motet.cli._api: api_request, normalize_base_url

Usage:
    motet-cli conversations list
    motet-cli conversations get <conversation_id>
    motet-cli conversations clear <conversation_id>
    motet-cli conversations rename <conversation_id> --title "New title"
    motet-cli conversations delete <conversation_id>
"""

from __future__ import annotations

import json
from typing import Optional
from urllib.parse import urlencode

import click

from ._api import api_request, api_url_option, normalize_base_url
from ._auth import get_api_headers

API_TIMEOUT = 60


@click.group("conversations")
def conversations_group() -> None:
    """Manage conversations (API: /api/v1/conversations)."""
    pass


@conversations_group.command("list")
@click.option("--agent-id", help="Filter by agent (e.g. default, motet_admin). Omitted defaults to core.default.")
@click.option("--surface-id", help="Filter by surface (e.g. demo_chat, ops_dashboard, cli). Omitted returns all surfaces for agent.")
@api_url_option()
def list_conversations(
    api_url: str,
    agent_id: Optional[str],
    surface_id: Optional[str],
) -> None:
    """List conversations (GET /api/v1/conversations). Optional agent_id and surface_id filters."""
    base = normalize_base_url(api_url)
    url = f"{base}/api/v1/conversations"
    q = {}
    if agent_id:
        q["agent_id"] = agent_id
    if surface_id:
        q["surface_id"] = surface_id
    if q:
        url = f"{url}?{urlencode(q)}"
    r = api_request("GET", url, headers=get_api_headers(), timeout=API_TIMEOUT)
    data = r.json()
    for c in data.get("conversations", []):
        parts = [c.get("id", ""), c.get("title", ""), f"updated={c.get('updated_at')}"]
        if c.get("agent_id"):
            parts.append(f"agent={c['agent_id']}")
        if c.get("surface_id"):
            parts.append(f"surface={c['surface_id']}")
        click.echo("  " + "  ".join(parts))


@conversations_group.command("get")
@click.argument("conversation_id")
@api_url_option()
def get_conversation(conversation_id: str, api_url: str) -> None:
    """Get conversation details (GET /api/v1/conversations/{id})."""
    base = normalize_base_url(api_url)
    r = api_request(
        "GET",
        f"{base}/api/v1/conversations/{conversation_id}",
        headers=get_api_headers(),
        timeout=API_TIMEOUT,
    )
    click.echo(json.dumps(r.json(), indent=2))


@conversations_group.command("clear")
@click.argument("conversation_id")
@api_url_option()
def clear_conversation(conversation_id: str, api_url: str) -> None:
    """Clear conversation (POST /api/v1/conversations/{id}/clear)."""
    base = normalize_base_url(api_url)
    r = api_request(
        "POST",
        f"{base}/api/v1/conversations/{conversation_id}/clear",
        headers=get_api_headers(),
        timeout=API_TIMEOUT,
    )
    data = r.json()
    click.echo(f"✅ Cleared: {data.get('cleared', {})}")


@conversations_group.command("rename")
@click.argument("conversation_id")
@click.option("--title", required=True, help="New title")
@api_url_option()
def rename_conversation(conversation_id: str, title: str, api_url: str) -> None:
    """Rename conversation (PATCH /api/v1/conversations/{id})."""
    base = normalize_base_url(api_url)
    r = api_request(
        "PATCH",
        f"{base}/api/v1/conversations/{conversation_id}",
        headers=get_api_headers(),
        json={"title": title},
        timeout=API_TIMEOUT,
    )
    click.echo("✅ Renamed.")


@conversations_group.command("delete")
@click.argument("conversation_id")
@api_url_option()
def delete_conversation(conversation_id: str, api_url: str) -> None:
    """Delete conversation (DELETE /api/v1/conversations/{id})."""
    base = normalize_base_url(api_url)
    api_request(
        "DELETE",
        f"{base}/api/v1/conversations/{conversation_id}",
        headers=get_api_headers(),
        timeout=API_TIMEOUT,
    )
    click.echo("✅ Deleted.")


__all__ = ["conversations_group"]

File: motet/interfaces/frontend/packages/motet-ui-common/src/api/conversations.ts
/**
 * Motet UI Common - Conversations API Client
 *
 * Copyright (c) 2024-2026 Motet Contributors
 * Licensed under the Apache License, Version 2.0. See LICENSE.
 *
 * Author: Matt Chisholm <matt@motet.dev>
 * Last Modified: 2026-08-31
 *
 * Description:
 *     Typed client for /api/v1/conversations (list, get, rename, delete).
 *     Pure fetch-based — no framework dependencies. mapHistoryToMessages
 *     groups consecutive attributed assistant rows into one turn.
 *
 * Usage:
 *     import { listConversations, getConversation } from "@motet/ui-common/api";
 */

import type { SpawnChildCard, ToolSummaryRow } from "../types";
import { groupTranscriptAssistantTurns } from "../utils/assistantTurn";
import { knownCostUsd } from "../utils/formatting";

// ─────────────────────────────────────────────────────────────────────────────
// RESPONSE TYPES
// ─────────────────────────────────────────────────────────────────────────────

export type ConversationItem = {
  id: string;
  title: string;
  created_at: number;
  updated_at: number;
  /** Agent scope (e.g. core.default) */
  agent_id?: string | null;
  /** Surface/channel (e.g. demo_chat, ops_dashboard) */
  surface_id?: string | null;
  /** Follow-up agent when it differs from list-scope agent_id */
  turn_agent_id?: string | null;
  /** Immediate parent, or null for a root chat */
  parent_conversation_id: string | null;
};

export type ConversationListResponse = {
  conversations: ConversationItem[];
};

export type ConversationHistoryAttachment = {
  artifact_id: string;
  content_type: string;
  filename?: string;
  bytes?: number;
};

export type ConversationHistoryItem = {
  content?: string;
  text?: string;
  role?: string;
  created_at?: string;
  /** Qualified agent id when present */
  agent_id?: string;
  /** Immediate parent agent when this row is a nested loop */
  parent_agent_id?: string;
  /** Provider reasoning stored for this assistant turn (reload display) */
  thinking_text?: string;
  /** Short tool name/status/preview for this assistant turn (reload display) */
  tool_summaries?: ToolSummaryRow[];
  /** Priced estimate for this assistant row; omitted when unknown */
  cost_usd?: number;
  /** Isolated spawn-child conversations created during this assistant turn */
  spawn_children?: SpawnChildCard[];
  /** Artifact references for media (e.g. images) so UI can display them */
  attachments?: ConversationHistoryAttachment[];
};

export type ConversationDetailResponse = {
  conversation_id: string;
  history: ConversationHistoryItem[];
  counts: { memory?: number; vector?: number };
  summary?: string | null;
  warning?: string | null;
  /** Priced conversation rollup; omitted when unknown */
  cost_usd?: number | null;
  /** Follow-up agent when it differs from the list-scope agent */
  turn_agent_id?: string | null;
  /** Immediate parent, or null for a root chat */
  parent_conversation_id: string | null;
};

// ─────────────────────────────────────────────────────────────────────────────
// REQUEST TYPES
// ─────────────────────────────────────────────────────────────────────────────

/** Optional scope for list (agent_id, surface_id). */
export type ListConversationsParams = {
  agent_id?: string | null;
  surface_id?: string | null;
};

// ─────────────────────────────────────────────────────────────────────────────
// API FUNCTIONS
// ─────────────────────────────────────────────────────────────────────────────

const BASE = "/api/v1/conversations";

/**
 * List conversations for the current principal.
 * Pass agent_id and surface_id to scope the list.
 * On 401, returns null so the app can still work (e.g. new chat).
 */
export async function listConversations(
  headers: Record<string, string>,
  params?: ListConversationsParams
): Promise<ConversationItem[] | null> {
  try {
    const search = new URLSearchParams();
    if (params?.agent_id != null && params.agent_id !== "") search.set("agent_id", params.agent_id);
    if (params?.surface_id != null && params.surface_id !== "") search.set("surface_id", params.surface_id);
    const url = search.toString() ? `${BASE}?${search.toString()}` : BASE;
    const r = await fetch(url, { headers });
    if (!r.ok) {
      if (r.status === 401) return null;
      throw new Error(`${r.status} ${r.statusText}`);
    }
    const data: ConversationListResponse = await r.json();
    return data.conversations ?? [];
  } catch {
    return null;
  }
}

/**
 * Get one conversation (history and metadata). Returns null on 404 or error.
 */
export async function getConversation(
  conversationId: string,
  headers: Record<string, string>
): Promise<ConversationDetailResponse | null> {
  try {
    const r = await fetch(`${BASE}/${conversationId}`, { headers });
    if (!r.ok) {
      if (r.status === 404) return null;
      throw new Error(`${r.status} ${r.statusText}`);
    }
    return (await r.json()) as ConversationDetailResponse;
  } catch {
    return null;
  }
}

/**
 * Rename a conversation (update display title). Throws on error.
 */
export async function updateConversationTitle(
  conversationId: string,
  title: string,
  headers: Record<string, string>
): Promise<{ conversation_id: string; title: string }> {
  const r = await fetch(`${BASE}/${conversationId}`, {
    method: "PATCH",
    headers: { ...headers, "Content-Type": "application/json" },
    body: JSON.stringify({ title: title.trim() }),
  });
  if (!r.ok) {
    const msg = (await r.json().catch(() => ({}))).detail ?? `${r.status} ${r.statusText}`;
    throw new Error(typeof msg === "string" ? msg : JSON.stringify(msg));
  }
  return (await r.json()) as { conversation_id: string; title: string };
}

/**
 * Delete a conversation (registry + memory/vector). Throws on error.
 */
export async function deleteConversation(
  conversationId: string,
  headers: Record<string, string>
): Promise<{ conversation_id: string; cleared: { memory?: number; vector?: number } }> {
  const r = await fetch(`${BASE}/${conversationId}`, {
    method: "DELETE",
    headers,
  });
  if (!r.ok) {
    const msg = (await r.json().catch(() => ({}))).detail ?? `${r.status} ${r.statusText}`;
    throw new Error(typeof msg === "string" ? msg : JSON.stringify(msg));
  }
  return (await r.json()) as { conversation_id: string; cleared: { memory?: number; vector?: number } };
}

/**
 * Map API history items to chat message shape (role, content, attachments).
 * Excludes tool/system messages so only user-facing transcript is returned.
 * Consecutive attributed assistant rows from one user turn are folded into a
 * single message with reconstructed agentStreams (selected agent, else last row).
 * thinking_text, tool_summaries, cost_usd, and spawn_children on a history item
 * are copied onto that agent's stream / the parent message.
 */
export function mapHistoryToMessages(
  history: ConversationHistoryItem[],
  selectedAgentId?: string
): Array<{
  role: "user" | "assistant";
  content: string;
  meta?: Record<string, unknown>;
  attachments?: Array<{
    artifact_id: string;
    content_type: string;
    filename?: string;
    bytes?: number;
    status: "ready";
  }>;
}> {
  const mapped = (history || [])
    .filter((m) => {
      const role = (m.role || "user") as string;
      if (role === "tool" || role === "system") return false;
      const thinking = (m.thinking_text || "").trim();
      const summaries = Array.isArray(m.tool_summaries) ? m.tool_summaries : [];
      const cost = knownCostUsd(m.cost_usd);
      const spawnChildren = Array.isArray(m.spawn_children) ? m.spawn_children : [];
      if (role === "assistant" && !(m.content ?? m.text ?? "").trim() && !thinking && summaries.length === 0 && cost == null && spawnChildren.length === 0) return false;
      return true;
    })
    .map((m) => {
      const attachments = m.attachments?.map((a) => ({
        artifact_id: a.artifact_id,
        content_type: a.content_type,
        filename: a.filename ?? "image",
        bytes: a.bytes ?? 0,
        status: "ready" as const,
      }));
      const thinking = (m.thinking_text || "").trim();
      const summaries = Array.isArray(m.tool_summaries) ? m.tool_summaries : [];
      const cost = knownCostUsd(m.cost_usd);
      const spawnChildren = Array.isArray(m.spawn_children) ? m.spawn_children : [];
      return {
        role: (m.role || "user") as "user" | "assistant",
        content: m.content ?? m.text ?? "",
        ...((m.agent_id || m.parent_agent_id || thinking || summaries.length || cost != null || spawnChildren.length)
          ? {
              meta: {
                ...(m.agent_id ? { agent_id: m.agent_id } : {}),
                ...(m.parent_agent_id ? { parent_agent_id: m.parent_agent_id } : {}),
                ...(thinking ? { thinking_text: thinking } : {}),
                ...(summaries.length ? { tool_summaries: summaries } : {}),
                ...(cost != null ? { cost_usd: cost } : {}),
                ...(spawnChildren.length ? { spawn_children: spawnChildren } : {}),
              },
            }
          : {}),
        ...(attachments?.length ? { attachments } : {}),
      };
    });
  return groupTranscriptAssistantTurns(mapped, selectedAgentId);
}
